File: src/main/cpp/subsystems/subsystem_LEDs.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "subsystems/subsystem_LEDs.h"
#include <frc/smartdashboard/SmartDashboard.h>

subsystem_LED::subsystem_LED(): m_CANdle{LEDConstants::CANdleID}, m_LEDTimer{} {

    m_LEDTimer.Start();
    
    m_CANdle.ConfigLEDType(ctre::phoenix::led::LEDStripType::RGB);
    m_CANdle.ConfigBrightnessScalar(0.5);
    m_CANdle.ConfigLOSBehavior(true);
}

LEDConstants::LEDState subsystem_LED::GetLEDState(){
    return m_LEDState;
}

void subsystem_LED::SetPurpleLED(){
    m_LEDState = LEDConstants::LEDState::Purple;
    m_TriggerState = LEDConstants::LEDState::Purple;
    // m_LEDTimer.Reset();
}

void subsystem_LED::SetYellowLED(){
    // m_LEDTimer.Reset();
    m_LEDState = LEDConstants::LEDState::Yellow;
    m_TriggerState = LEDConstants::LEDState::Yellow;

}

void subsystem_LED::SetStandbyLED(){
    m_LEDState = LEDConstants::LEDState::Standby;

}

void subsystem_LED::ToggleFlashLEDs(){
    flashLEDs = !flashLEDs;
}

void subsystem_LED::SetErrorLED(){
    m_LEDState = LEDConstants::LEDState::Error;

}

void subsystem_LED::SetOffLED(){
    m_LEDState = LEDConstants::LEDState::Off;
}

void subsystem_LED::SetTriggerStateLED(){
    m_LEDState = m_TriggerState;
}

void subsystem_LED::cycleLED(){
    switch (m_LEDState){
        case(LEDConstants::LEDState::Purple):
            SetYellowLED();
            frc::SmartDashboard::SmartDashboard::PutString("LED Color", "Yellow");
            break;
        default:
            SetPurpleLED();
            frc::SmartDashboard::SmartDashboard::PutString("LED Color", "Purple");
    }
}

// This method will be called once per scheduler run
void subsystem_LED::Periodic() {
    if (flashLEDs && std::fmod(m_LEDTimer.Get().value(), 10) > 5 ){
        m_CANdle.SetLEDs(255,255,255);
    } else {
        switch(m_LEDState){
            case(LEDConstants::LEDState::Standby):
            // green red blue
                m_CANdle.SetLEDs(255, 0, 0);
                break;
            case(LEDConstants::LEDState::Purple):
                m_CANdle.SetLEDs(0, 255, 255);
                break;
            case(LEDConstants::LEDState::Yellow):
                m_CANdle.SetLEDs(255, 255, 0);
                break;
            case(LEDConstants::LEDState::Error):
                m_CANdle.SetLEDs(0, 255, 0);
                break;
            case(LEDConstants::LEDState::Off):
                m_CANdle.SetLEDs(0, 0, 255);
                break;
        }
    }
       
}

frc2::CommandPtr subsystem_LED::cycleLEDCommand(){
    return RunOnce([this]{return cycleLED();});
}

File: src/main/java/frc/robot/subsystems/subsystem_LEDs.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems;

import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj2.command.CommandBase;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import frc.robot.Constants.LEDConstants;
import frc.robot.Constants.LEDConstants.LEDState;

import com.ctre.phoenix.led.CANdle;
//import com.ctre.phoenix.led.ColorFlowAnimation;
import com.ctre.phoenix.led.SingleFadeAnimation;
//import com.ctre.phoenix.led.StrobeAnimation;
import com.ctre.phoenix.led.CANdle.LEDStripType;

public class subsystem_LEDs extends SubsystemBase {
  /** Creates a new subsystem_LEDs. */
  private CANdle m_CANdle;
  private Timer m_LEDTimer;
  private boolean m_FlashLEDs;
  private SingleFadeAnimation m_IntakedAnimation;
  private LEDState m_LEDState;
  private LEDState m_TriggerState;
  public subsystem_LEDs() {
    m_CANdle = new CANdle(LEDConstants.candleID);
    m_LEDTimer = new Timer();
    m_FlashLEDs = false;
    m_IntakedAnimation = new SingleFadeAnimation(0, 255, 0, 0, 0.5, 40);
    m_LEDState = LEDState.standby;
    m_TriggerState = LEDState.standby;

    m_LEDTimer.start();
    m_CANdle.configLEDType(LEDStripType.RGB);
    m_CANdle.configBrightnessScalar(0.5);
    m_CANdle.configLOSBehavior(true);
  }

  public LEDState getLEDState(){
    return m_LEDState;
  }

  public void setPurpleLED(){
    m_LEDState = LEDState.purple;
    m_TriggerState = LEDState.purple;
  }

  public void setYellowLED(){
    m_LEDState = LEDState.yellow;
    m_TriggerState = LEDState.yellow;
  }

  public void setStandbyLED(){
    m_LEDState = LEDState.standby;
  }

  public void toggleFlashLEDs(){
    m_FlashLEDs = !m_FlashLEDs;
  }

  public void setErrorLED(){
    m_LEDState = LEDState.error;
  }

  public void setOffLED(){
    m_LEDState = LEDState.off;
  }

  public void setTriggerStateLED(){
    m_LEDState = m_TriggerState;
  }

  public void cycleLED(){
    switch (m_LEDState){
      case purple:
        setYellowLED();
        break;
      default:
        setPurpleLED();
        break;
    }
  }

  @Override
  public void periodic() {
    // This method will be called once per scheduler run
    if (m_FlashLEDs && m_LEDTimer.get() % 10 > 5.0){
      m_CANdle.setLEDs(255, 255, 255);
    } else {
      switch(m_LEDState){
        case standby:
          m_CANdle.setLEDs(0, 255, 0);
          break;
        case purple:
          m_CANdle.setLEDs(255, 0, 255);
          break;
        case yellow:
          m_CANdle.setLEDs(255, 255, 0);
          break;
        case off:
          m_CANdle.setLEDs(0, 0, 255);
          break;
        case error:
        default:
          m_CANdle.setLEDs(0, 255, 0);
          break;
      }
    }
  }

  public CommandBase cycleLEDCommand(){
      // implicitly require `this`
      return this.runOnce(() -> cycleLED());
  }

}
